pulp_ansible/app/models.py

```diff
@@ -169,7 +172,7 @@ def relative_path(self):
 
     class Meta:
         default_related_name = "%(app_label)s_%(model_name)s"
-        unique_together = ("namespace", "name", "version")
+        unique_together = ("sha256",)
```

Comment:
So the artifact contains the information about namespace, name and version.
Do you see any conceivable way (maybe through on-demand sync) that pulp could be fooled about that information? Is this a problem we intend to solve, or do we in general trust upstream sources?

Reply:
Which information being fooled? The name, namespace, version or the sha256? name, namespace, version is up to the user so it could be what ever they want. If you are talking about it differing from the downloaded artifact, then yes it is possible if an upstream messes with the database to change the metadata served or change which artifact is pointed to. I think in general we trust the upstream source (and since upstream will either be galaxy or pulp I think this is a safe assumption to make. Maybe this something we can consider in #1106 .

Reply:
Is this the `sha256` of the tar.gz itself, or the `sha256` of the `MANIFEST.json`?

Since `ansible-galaxy collection build` always generates a new `MANIFEST.json` and `FILES.json`, we set the `mtime` to `time.time()` for those files in the tarball.  So if you simply run `ansible-galaxy collection build` 2 times, the `tar.gz` will have different `sha256`

Would this cause problems for you if someone tried to upload a rebuilt artifact?

As it is, the `sha256` of the tarball isn't necessarily equivalent of `(namepace, name, version)`.

Reply:
This new `sha256` field is for the tarball `sha256` (the underlying artifact of the CollectionVersion).

Uploading a rebuilt artifact wouldn't break Pulp, it would just cause confusion for the user. With this change Pulp would gladly accept both the original and rebuilt collection since they would have different digests. Both could be stored in Pulp, but the repository you have them in would only hold the latest uploaded one. Doing a global content list for `(namespace, name, version)` could now return multiple results so the user would have to know the date uploaded or `sha256` to select the right one. 

However, having the new field be the `sha256` of the `MANIFEST.json` might be a better idea. In that scenario would the `sha256` of the `MANIFEST.json` of the second built artifact be the same as the first?

Reply:
The `MANIFEST.json` will have the same sha, unless of course if something else was changed:

```console
$ shasum -a 256 sivel-toiletwater-0.0.10.tar.gz*
878311263fcd21f040022a33187d682a63a798e16a79d9bdf91d65e8af30b1a6  sivel-toiletwater-0.0.10.tar.gz.1
5fb2e8026e5dae65e1c40c19fa168883c524f3f99c90ec67a58ebe6ea920b51f  sivel-toiletwater-0.0.10.tar.gz.2
$ tar -Oxzvf sivel-toiletwater-0.0.10.tar.gz.1 MANIFEST.json | shasum -a 256
x MANIFEST.json
9a0812ebbd0107139c0e3e4684d585393cfce6a8d9fff4876e6cabd3e244ade7  -
$ tar -Oxzvf sivel-toiletwater-0.0.10.tar.gz.2 MANIFEST.json | shasum -a 256
x MANIFEST.json
9a0812ebbd0107139c0e3e4684d585393cfce6a8d9fff4876e6cabd3e244ade7  -
```

Reply:
Thank you for testing this. Am I safe to assume that the `MANIFEST.json` is source of truth for two collections being the same? If so, I'll change the new field to be the `MANIFEST.json`'s `sha256` to remove the issue of uploading duplicate build artifacts.

Reply:
I guess that is up for interpretation, but to clarify, the `MANIFEST.json` holds a `sha256` for `FILES.json`, which holds a `sha256` for every file in the artifact.  So the `sha256` of a `MANIFEST.json` which also holds `namespace`, `name`, and `version` should accurately represent whether 2 collections are different.

Reply:
My question is whether the FILES.json is always ordered in the same way, thereby giving the same checksum if the same files are present.

Reply:
The order of `FILES.json` is determined by filesystem order, effectively that of `os.listdir`.